File: Sources/RepoBarCore/Models/ProviderCapabilities.swift
import Foundation

public struct ProviderCapabilities: Equatable, Codable, Sendable {
    public var repositories: Bool
    public var issues: Bool
    public var pullRequests: Bool
    public var ci: Bool
    public var releases: Bool
    public var tags: Bool
    public var branches: Bool
    public var commits: Bool
    public var contributors: Bool
    public var repositoryContents: Bool
    public var contributionCalendar: Bool
    public var trafficStats: Bool
    public var discussions: Bool
    public var rateLimitDiagnostics: Bool

    public init(
        repositories: Bool,
        issues: Bool,
        pullRequests: Bool,
        ci: Bool,
        releases: Bool,
        tags: Bool,
        branches: Bool,
        commits: Bool,
        contributors: Bool,
        repositoryContents: Bool,
        contributionCalendar: Bool,
        trafficStats: Bool,
        discussions: Bool,
        rateLimitDiagnostics: Bool
    ) {
        self.repositories = repositories
        self.issues = issues
        self.pullRequests = pullRequests
        self.ci = ci
        self.releases = releases
        self.tags = tags
        self.branches = branches
        self.commits = commits
        self.contributors = contributors
        self.repositoryContents = repositoryContents
        self.contributionCalendar = contributionCalendar
        self.trafficStats = trafficStats
        self.discussions = discussions
        self.rateLimitDiagnostics = rateLimitDiagnostics
    }

    public static let github = ProviderCapabilities(
        repositories: true,
        issues: true,
        pullRequests: true,
        ci: true,
        releases: true,
        tags: true,
        branches: true,
        commits: true,
        contributors: true,
        repositoryContents: true,
        contributionCalendar: true,
        trafficStats: true,
        discussions: true,
        rateLimitDiagnostics: true
    )

    public static let gitlab = ProviderCapabilities(
        repositories: true,
        issues: true,
        pullRequests: true,
        ci: true,
        releases: true,
        tags: true,
        branches: true,
        commits: true,
        contributors: true,
        repositoryContents: true,
        contributionCalendar: false,
        trafficStats: false,
        discussions: false,
        rateLimitDiagnostics: true
    )

    public static let bitbucketCloud = ProviderCapabilities(
        repositories: true,
        issues: true,
        pullRequests: true,
        ci: true,
        releases: false,
        tags: true,
        branches: true,
        commits: true,
        contributors: false,
        repositoryContents: true,
        contributionCalendar: false,
        trafficStats: false,
        discussions: false,
        rateLimitDiagnostics: true
    )

    public static let forgejoCompatible = ProviderCapabilities(
        repositories: true,
        issues: true,
        pullRequests: true,
        ci: false,
        releases: true,
        tags: true,
        branches: true,
        commits: true,
        contributors: true,
        repositoryContents: true,
        contributionCalendar: false,
        trafficStats: false,
        discussions: false,
        rateLimitDiagnostics: true
    )

    public static let customGit = ProviderCapabilities(
        repositories: true,
        issues: false,
        pullRequests: false,
        ci: false,
        releases: false,
        tags: false,
        branches: true,
        commits: true,
        contributors: false,
        repositoryContents: false,
        contributionCalendar: false,
        trafficStats: false,
        discussions: false,
        rateLimitDiagnostics: false
    )
}

public extension SourceControlProvider {
    var defaultCapabilities: ProviderCapabilities {
        switch self {
        case .github:
            .github
        case .gitlab:
            .gitlab
        case .bitbucketCloud:
            .bitbucketCloud
        case .forgejo, .gitea:
            .forgejoCompatible
        case .customGit:
            .customGit
        }
    }
}

File: Windows/RepoBar.Core/Models/ProviderCapabilities.cs
namespace RepoBar.Core.Models;

public sealed record ProviderCapabilities(
    bool Repositories,
    bool Issues,
    bool PullRequests,
    bool Ci,
    bool Releases,
    bool Tags,
    bool Branches,
    bool Commits,
    bool Contributors,
    bool RepositoryContents,
    bool ContributionCalendar,
    bool TrafficStats,
    bool Discussions,
    bool RateLimitDiagnostics)
{
    public static ProviderCapabilities GitHub { get; } = new(
        Repositories: true,
        Issues: true,
        PullRequests: true,
        Ci: true,
        Releases: true,
        Tags: true,
        Branches: true,
        Commits: true,
        Contributors: true,
        RepositoryContents: true,
        ContributionCalendar: true,
        TrafficStats: true,
        Discussions: true,
        RateLimitDiagnostics: true);

    public static ProviderCapabilities GitLab { get; } = new(
        Repositories: true,
        Issues: true,
        PullRequests: true,
        Ci: true,
        Releases: true,
        Tags: true,
        Branches: true,
        Commits: true,
        Contributors: true,
        RepositoryContents: true,
        ContributionCalendar: false,
        TrafficStats: false,
        Discussions: false,
        RateLimitDiagnostics: true);

    public static ProviderCapabilities BitbucketCloud { get; } = new(
        Repositories: true,
        Issues: true,
        PullRequests: true,
        Ci: true,
        Releases: false,
        Tags: true,
        Branches: true,
        Commits: true,
        Contributors: false,
        RepositoryContents: true,
        ContributionCalendar: false,
        TrafficStats: false,
        Discussions: false,
        RateLimitDiagnostics: true);

    public static ProviderCapabilities ForgejoCompatible { get; } = new(
        Repositories: true,
        Issues: true,
        PullRequests: true,
        Ci: false,
        Releases: true,
        Tags: true,
        Branches: true,
        Commits: true,
        Contributors: true,
        RepositoryContents: true,
        ContributionCalendar: false,
        TrafficStats: false,
        Discussions: false,
        RateLimitDiagnostics: true);

    public static ProviderCapabilities CustomGit { get; } = new(
        Repositories: true,
        Issues: false,
        PullRequests: false,
        Ci: false,
        Releases: false,
        Tags: false,
        Branches: true,
        Commits: true,
        Contributors: false,
        RepositoryContents: false,
        ContributionCalendar: false,
        TrafficStats: false,
        Discussions: false,
        RateLimitDiagnostics: false);

    public static ProviderCapabilities For(SourceControlProvider provider) =>
        provider switch
        {
            SourceControlProvider.GitHub => GitHub,
            SourceControlProvider.GitLab => GitLab,
            SourceControlProvider.BitbucketCloud => BitbucketCloud,
            SourceControlProvider.Forgejo or SourceControlProvider.Gitea => ForgejoCompatible,
            SourceControlProvider.CustomGit => CustomGit,
            _ => CustomGit,
        };
}
